WalletWasabi/WabiSabi/Models/MultipartyTransaction/Signing.cs

```diff
@@ -0,0 +1,88 @@
+using NBitcoin;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using WalletWasabi.WabiSabi.Backend.Models;
+
+namespace WalletWasabi.WabiSabi.Models.MultipartyTransaction
+{
+	public record Signing(Parameters Parameters, ImmutableArray<Coin> Inputs, ImmutableArray<TxOut> Outputs) : State
+	{
+		public ImmutableDictionary<int, WitScript> Witnesses { get; init; } = ImmutableDictionary<int, WitScript>.Empty;
+
+		public bool IsFullySigned => Witnesses.Count() == Inputs.Length;
+
+		public FeeRate EffectiveFeeRate => new FeeRate(Balance, EstimatedVsize);
+
+		public Money Balance => Inputs.Sum(x => x.Amount) - Outputs.Sum(x => x.Value);
+
+		public int EstimatedVsize => Parameters.SharedOverhead + EstimatedInputsVsize + OutputsVsize;
+		public int EstimatedInputsVsize => Inputs.Sum(x => x.TxOut.ScriptPubKey.EstimateInputVsize());
+		public int OutputsVsize => Outputs.Sum(x => x.ScriptPubKey.EstimateOutputVsize());
+
+		public IEnumerable<Coin> UnsignedInputs => Enumerable.Range(0, Inputs.Length).Where(i => !IsInputSigned(i)).Select(i => Inputs[i]);
+
+		public bool IsInputSigned(int index) => Witnesses.ContainsKey(index);
+
+
+		public Signing AddWitness(int index, WitScript witness)
+		{
+			if (IsInputSigned(index))
+			{
+				throw new WabiSabiProtocolException(WabiSabiProtocolErrorCode.WitnessAlreadyProvided);
+			}
+
+			// Verify witness.
+			// 1. Copy UnsignedCoinJoin.
+			Transaction cjCopy = CreateUnsignedTransaction();
+
+			// 2. Sign the copy.
+			cjCopy.Inputs[index].WitScript = witness;
+
+			// 3. Convert the current input to IndexedTxIn.
+			IndexedTxIn currentIndexedInput = cjCopy.Inputs.AsIndexedInputs().Skip(index).First();
```

Comment:
This looks so ugly, If we create the NBitcoinExtension method we should add one for this:

```c#
			IndexedTxIn currentIndexedInput =  cjCopy.Inputs.GetIndexedInput(index);
```

The method is something like:
```c#
public static IndexedTxIn GetIndexedInput(this TxInList list, uint i) =>
	new IndexedTxIn {
		TxIn = list.Transaction.Inputs[i],
		Index = i,
		Transaction = list.Transaction
	};
```

It could be a local static method too but i am not so sure if that is better or worse in this case.

Reply:
I think this is a useful extension. I didn't modify this code, only moved it from `Arena`, I think we can do this as a separate cleanup PR and I can rebase?